msgraph-developer-proxy/ProxyEngine.cs

```diff
@@ -95,11 +98,80 @@ public class ProxyEngine {
         _logger.LogInfo("Press CTRL+C to stop the Microsoft Graph Developer Proxy");
         _logger.LogInfo("");
         Console.CancelKeyPress += Console_CancelKeyPress;
-        // wait for the proxy to stop
-        Console.ReadLine();
+
+        if (_config.Record) {
+            StartRecording();
+        }
+        _pluginEvents.AfterRequestLog += AfterRequestLog;
+
+        // we need this check or proxy will fail with an exception
+        // when run for example in VSCode's integrated terminal
+        if (!Console.IsInputRedirected) {
+            ReadKeys();
+        }
         while (_proxyServer.ProxyRunning) { Thread.Sleep(10); }
     }
 
+    private void AfterRequestLog(object? sender, RequestLogArgs e) {
+        if (!_isRecording)
+        {
+            return;
+        }
+
+        _requestLogs.Add(e.RequestLog);
+    }
+
+    private void ReadKeys() {
+        ConsoleKey key;
+        do {
+            key = Console.ReadKey(true).Key;
+            if (key == ConsoleKey.R) {
+                StartRecording();
+            }
+            else if (key == ConsoleKey.S) {
+                StopRecording();
+            }
+        } while (key != ConsoleKey.Escape);
+    }
+
+    private void StartRecording() {
+        if (_isRecording) {
+            return;
+        }
+
+        _isRecording = true;
+        PrintRecordingIndicator();
+    }
+
+    private void StopRecording() {
+        if (!_isRecording) {
+            return;
+        }
+
+        _isRecording = false;
+        PrintRecordingIndicator();
+        // clone the list so that we can clear the original
+        // list in case a new recording is started, and
+        // we let plugins handle previously recorded requests
+        var clonedLogs = _requestLogs.ToArray();
+        _requestLogs.Clear();
+        _pluginEvents.RaiseRecordingStopped(new RecordingArgs(clonedLogs));
+    }
+
+    private void PrintRecordingIndicator() {
+        lock (ConsoleLogger.ConsoleLock) {
+            if (_isRecording) {
+                Console.ForegroundColor = ConsoleColor.Red;
+                Console.Error.Write("◉");
```

Comment:
I looked into making the indicator fixed in the terminal window but it turned out to be too complex for the current setup with continuous logging, without using any TUI framework. For now, the indicator is logged one-time to stderr.